File: lib/domain/user/user_service.dart
// Copyright © 2020 WorldRIZe. All rights reserved.

import 'package:data_classes/data_classes.dart';
import 'package:wr_app/domain/auth/auth_repository.dart';
import 'package:wr_app/domain/user/model/membership.dart';
import 'package:wr_app/domain/user/model/user.dart';
import 'package:wr_app/domain/user/model/user_api_dto.dart';
import 'package:wr_app/domain/user/user_repository.dart';

// TODO(someone): Error handling
// update api returns updated object
class UserService {
  const UserService({
    @required IAuthRepository authRepository,
    @required IUserRepository userRepository,
  })  : _userRepository = userRepository,
        _authRepository = authRepository;

  final IAuthRepository _authRepository;
  final IUserRepository _userRepository;

  /// sign up with Google
  Future<User> signUpWithGoogle() async {
    await _authRepository.signInWithGoogleSignIn();
    final _res = await _userRepository.createUser(CreateUserRequestDto());
    return _res.user;
  }

  /// sign up with email & password
  Future<User> signUpWithEmailAndPassword(String email, String password) async {
    await _authRepository.signUpWithEmailAndPassword(email, password);
    final _res = await _userRepository.createUser(CreateUserRequestDto());
    return _res.user;
  }

  /// sign in with email & password
  Future<User> signInWithEmailAndPassword(String email, String password) async {
    await _authRepository.signInWithEmailAndPassword(email, password);
    final _res = await _userRepository.readUser(ReadUserRequestDto());
    return _res.user;
  }

  /// sign out
  Future<void> signOut() {
    return _authRepository.signOut();
  }

  /// call test api
  Future<void> test() async {
    return _userRepository.test(TestRequestDto());
  }

  /// ユーザーデータを習得します
  Future<User> getUser() async {
    final _res = await _userRepository.readUser(ReadUserRequestDto());
    return _res.user;
  }

  /// フレーズをお気に入りに登録します
  Future<User> favorite({
    @required User user,
    @required String phraseId,
    @required bool value,
  }) async {
    final res = await _userRepository.favoritePhrase(FavoritePhraseRequestDto(
        uid: user.uuid, phraseId: phraseId, value: value));

    if (!(res?.success ?? false)) {
      throw Exception('failed');
    }

    user.favorites[phraseId] = value;
    return user;
  }

  /// 受講可能回数をリセット
  Future<User> resetTestCount({
    @required User user,
  }) async {
    user.testLimitCount = 3;
    final res =
        await _userRepository.updateUser(UpdateUserRequestDto(user: user));

    if (!(res?.success ?? false)) {
      throw Exception('failed');
    }

    return user;
  }

  /// ポイントを習得します
  Future<User> getPoints({
    @required User user,
    @required int point,
  }) async {
    user.point += point;
    final res = await _userRepository.getPoint(GetPointRequestDto(
      uid: user.uuid,
      point: point,
    ));

    if (!(res?.success ?? false)) {
      throw Exception('failed');
    }

    return user;
  }

  Future<User> updateAge({@required User user, @required String age}) async {
    user.age = age;
    final res =
        await _userRepository.updateUser(UpdateUserRequestDto(user: user));

    if (!(res?.success ?? false)) {
      throw Exception('failed');
    }

    return user;
  }

  /// upgrade to Pro or downgrade
  Future<User> changePlan(
      {@required User user, @required Membership membership}) async {
    user.membership = membership;

    final res =
        await _userRepository.updateUser(UpdateUserRequestDto(user: user));

    if (!(res?.success ?? false)) {
      throw Exception('failed');
    }

    return user;
  }

  /// do test
  Future<User> doTest({@required User user}) async {
    user.testLimitCount--;
    final res = await _userRepository.doTest(DoTestRequestDto(uid: user.uuid));

    if (!(res?.success ?? false)) {
      throw Exception('failed');
    }

    return user;
  }

  /// update user
  Future<User> updateUser({@required User user}) async {
    final res =
        await _userRepository.updateUser(UpdateUserRequestDto(user: user));

    if (!(res?.success ?? false)) {
      throw Exception('failed');
    }

    return user;
  }

  /// update password
  Future<void> updatePassword(
      {@required String currentPassword, @required String newPassword}) {
    return _authRepository.updatePassword(currentPassword, newPassword);
  }

  /// update email
  Future<User> updateEmail(
      {@required User user, @required String newEmail}) async {
    user.email = newEmail;

    await _authRepository.updateEmail(newEmail);

    return user;
  }
}

File: functions/src/domain/user/userService.ts
/**
 * Copyright © 2020 WorldRIZe. All rights reserved.
 */
import { User } from './model/user'
import { CreateUserRequest, UpdateUserRequest } from './model/userApiDto'
import { UserRepository } from './userRepository'

export class UserService {
  private readonly repo: UserRepository

  constructor(repo: UserRepository) {
    this.repo = repo
  }

  generateUser (uuid: string, req: CreateUserRequest): User {
    return {
      uuid: uuid,
      name: req.name,
      userId: uuid,
      point: 0,
      testLimitCount: 3,
      favorites: {},
      email: req.email,
      age: req.age,
    }
  }

  async existUser(uuid: string): Promise<boolean> {
    return this.repo.exists(uuid)
  }

  async updateUser(req: UpdateUserRequest): Promise<User> {
    return this.repo.update(req.user)
  }

  async createUser(uuid: string, req: CreateUserRequest) {
    const user = this.generateUser(uuid, req)
    return this.repo.create(user)
  }

  async readUser(uuid: string): Promise<User | null> {
    return this.repo.findById(uuid)
  }

  async favoritePhrase (uuid: string, phraseId: string, value: boolean): Promise<boolean> {
    const user = await this.repo.findById(uuid)
    if (!user) {
      throw 'User not found'
    }
    user.favorites[phraseId] = value
    await this.repo.update(user)
    return true
  }

  async getPoint (uuid: string, point: number): Promise<boolean> {
    const user = await this.repo.findById(uuid)
    if (!user) {
      throw 'User not found'
    }
    if (user.point + point < 0) {
      throw 'Not enough points'
    }
    user.point += point
    await this.repo.update(user)
    return true
  }

  async delete(uuid: string): Promise<void> {
    return await this.repo.remove(uuid)
  }

  async doTest(uuid: string): Promise<boolean> {
    const user = await this.repo.findById(uuid)
    if (!user) {
      throw 'User not found'
    }
    if (user.testLimitCount == 0) {
      throw 'Daily test limit exceeded'
    }
    user.testLimitCount -= 1
    await this.repo.update(user)
    return true
  }
}
